src/coord/src/session.rs

```diff
@@ -356,8 +380,10 @@ pub enum TransactionOps {
     /// The transaction has been initiated, but no statement has yet been executed
     /// in it.
     None,
-    /// This transaction has had a read (`SELECT`, `TAIL`) and must only do other reads.
-    Reads,
+    /// This transaction has had a peek (`SELECT`, `TAIL`) and must only do other peeks.
+    Peeks(Timestamp),
+    /// This transaction has done a TAIL and must do nothing else.
+    Tail,
```

Comment:
The following error is back for tail-only transactions over materialized views:

```
Dataflow tail-source-materialize.public.v1 requested as_of (Antichain { elements: [1618903297300] }) not >= since (Antichain { elements: [1618903297301] })
```
The two numbers always differ by exactly 1 ms.

To reproduce, run this test in a loop with --workers 1 . It should fail within 10-20 executions with the error above:

```
> CREATE TABLE t1 (f1 INTEGER);

> INSERT INTO t1 VALUES (1);

> INSERT INTO t1 VALUES (1);

> INSERT INTO t1 VALUES (1);

> CREATE MATERIALIZED VIEW v1 AS SELECT * FROM t1;

> BEGIN;

> DECLARE c1 CURSOR FOR TAIL v1;

> FETCH 1 c1;
<TIMESTAMP> 3 1
```

Reply:
Running this in a loop with 1 worker, the most recent code doesn't seem to trigger it?

Reply:
(I fixed a bug found in another test that may have also fixed this.)